### ISSUE TITLE
[Bug]: file chooser makes it appear as though apps have unrestricted filesystem access

### ISSUE THREAD

**Issue by alidehghan** [flatpak/flatpak#5880]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.14.8

### What Linux distribution are you using?

Ubuntu

### Linux distribution version

24

### What architecture are you using?

x86_64

### How to reproduce

flatpak run --nofilesystem=host --nofilesystem=home --user com.google.Chrome

### Expected Behavior

when I run chrome with (flatpak run --nofilesystem=host --nofilesystem=home --user com.google.Chrome), I expect not to expose home or desktop or other filesystems to the running chrome, whereas when right click on an image (in any website browsed in sandboxed chrome), I can view (and even save) the image into any filesystem available !!!

### Actual Behavior

hide directories and prevent from saving images or everything (download as) into filesystems.

**Comment by hfiguiere** [comment]
Given the filesystem permissions of this speccific package this behave as intended.

From the `flatpak-run` manpage:

>        --filesystem=FILESYSTEM
>            Allow the application access to a subset of the filesystem. This overrides to the Context section from the application metadata.  FILESYSTEM
>            can be one of: home, host, host-os, host-etc, xdg-desktop, xdg-documents, xdg-download, xdg-music, xdg-pictures, xdg-public-share,
>            xdg-templates, xdg-videos, xdg-run, xdg-config, xdg-cache, xdg-data, an absolute path, or a homedir-relative path like ~/dir or paths
>            relative to the xdg dirs, like xdg-download/subdir. The optional :ro suffix indicates that the location will be read-only. The optional
>            :create suffix indicates that the location will be read-write and created if it doesn't exist. This option can be used multiple times. See
>            the "[Context] filesystems" list in flatpak-metadata(5) for details of the meanings of these filesystems.
> 
>        --nofilesystem=FILESYSTEM
>            Undo the effect of a previous --filesystem=FILESYSTEM in the app's manifest and/or the overrides set up with flatpak-override(1). This
>            overrides the Context section of the application metadata.  FILESYSTEM can take the same values as for --filesystem, but the :ro and :create
>            suffixes are not used here. This option can be used multiple times.
> 
>            This option does not prevent access to a more narrowly-scoped --filesystem. For example, if an application has the equivalent of
>            --filesystem=xdg-config/MyApp in its manifest or as a system-wide override, and flatpak override --user --nofilesystem=home as a per-user
>            override, then it will be prevented from accessing most of the home directory, but it will still be allowed to access $XDG_CONFIG_HOME/MyApp.
> 
>            As a special case, --nofilesystem=host:reset will ignore all --filesystem permissions inherited from the app manifest or flatpak-override(1),
>            in addition to having the behaviour of --nofilesystem=host.

**Comment by Mikenux** [comment]
Also note that if you use the correct overrides but Chrome uses the file chooser portal, you will still see that you can save an image anywhere when the file chooser is shown. If this is the case, be aware that this does not mean that Chrome is able to see the filesystem.

**Comment by alidehghan** [comment]
I have overrided setting and tested nofilesystem noted as above, but Chrome still is able to see, enumerate and save the file whenever user wants. In fact, no sandboxing is done in terms of FileSystem

**Comment by Mikenux** [comment]
How you checked that it is able to see, enumerate and save the file?

**Comment by alidehghan** [comment]
through the dialog, and saving successfully on the Desktop of the user

**Comment by Mikenux** [comment]
This is probably because Chrome uses the file chooser portal, which also handles file saving. Overriding static permissions does not disable portals. Portals are external processes that the app does not have access to, so when the file chooser (which in this case is a "file saver") is presented, Chrome cannot see your files. Chrome will only have access to the file you save.

For reference:
- https://github.com/flatpak/flatpak/issues/3977#issuecomment-1028979335
- Similar issue: https://github.com/flatpak/flatpak/issues/5391

If you are concerned about whether or not the file chooser/save you are facing allows an app to see or access your files, there are the following open issues (which are the same):
- https://github.com/flatpak/xdg-desktop-portal-gtk/issues/429
- https://gitlab.gnome.org/Teams/Design/os-mockups/-/issues/221

**Comment by alidehghan** [comment]
> Chrome cannot see your files

Yes it can see.

**Comment by alidehghan** [comment]
Inability to restrict accessing filesystems is a huge drawback, and I should verify that Flatpak couldn't solve this major challenge even after two years. This makes Flatpak very weak in terms of isolation/sandboxing. Hope they can manage it.

**Comment by Mikenux** [comment]
The file chooser can see your files but this file chooser does not belongs to the Chrome process. The file chooser portal just allow users to control which files the app has access to - so this requires user input. As such and if you removed the filesystem permissions, Chrome cannot see your files. The files Chrome can see and access are only those you saved from it. So, no Flatpak is not "very weaks in terms of isolation/sandboxing".

Disabling portals (as you commented on an issue I mentioned) is another thing.

**Comment by alidehghan** [comment]
Would you please give me a sample (cmd) to do that? I have tried many options over and over, totally in vein.

**Comment by rusty-snake** [comment]
> Yes it can see.

Proof missing.

> Would you please give me a sample (cmd) to do that?

- `flatpak run --command=sh <APP-ID>` inspect the filesystem.
- `flatpak run --no-documents-portal <APP-ID>` run w/o documents portal [untested]

**Comment by alidehghan** [comment]
> > Yes it can see.
> 
> Proof missing.
> 
> > Would you please give me a sample (cmd) to do that?
> 
> * `flatpak run --command=sh <APP-ID>` inspect the filesystem.
> * `flatpak run --no-documents-portal <APP-ID>` run w/o documents portal [untested]

![image](https://github.com/user-attachments/assets/7e7455a0-af07-4e06-ae3c-4fa685f720c9)

**Comment by smcv** [comment]
Your screenshot demonstrates that the file chooser, which runs outside the sandbox, can access the filesystem.

But, if you try to open or save a (harmless, expendable) file while running with `--no-documents-portal`, that should fail.

**Comment by alidehghan** [comment]
yes, I need to File Chooser dialog inside sandbox, is that possible?

**Comment by smcv** [comment]
Not without altering the app. The entire point of the file chooser portal is that it gives you control over what files the sandboxed app can and can't open, without having to make all of those decisions in advance.

**Comment by smcv** [comment]
To clarify, the partially-trusted sandboxed app *does not* have access to the full filesystem (unless it's given that access with `--filesystem=host` or similar).

It is only the trusted file chooser dialog that has unrestricted access. That dialog is part of an xdg-desktop-portal backend (probably xdg-desktop-portal-gnome or -gtk in your case). When you choose a file to open or to save as, the file chooser instructs the documents portal (xdg-documents-portal, part of xdg-desktop-portal) to give the app access to that single file. You can see the accesses that have been granted in this way by running `flatpak permissions`.

For example, after loading a file into GNOME Recipes, among other permissions I can see:

```
$ flatpak permissions
Table     Object              App                  Permissions                  Data
documents 225cdcb3            org.gnome.Recipes    read,write,grant-permissions (b'/home/smcv/tmp/emblem-debian.png', 43, 1, 0)
```

**Comment by smcv** [comment]
See also https://github.com/flatpak/xdg-desktop-portal-gtk/issues/429

**Comment by daniliammo** [comment]
@alidehghan 
> Inability to restrict accessing filesystems is a huge drawback, and I should verify that Flatpak couldn't solve this major challenge even after two years. This makes Flatpak very weak in terms of isolation/sandboxing. Hope they can manage it.

Flatpak can deny access to file systems. Try running the flatseal app to easily set up app permissions

**Comment by chrisawi** [comment]
Closing in favor of https://github.com/flatpak/flatpak/issues/3977 and https://github.com/flatpak/xdg-desktop-portal-gtk/issues/429
